runtime/minor_gc.c

```diff
@@ -654,6 +654,8 @@ void caml_empty_minor_heap_promote(caml_domain_state* domain,
                   Bsize_wsize(domain->allocated_words - prev_alloc_words));
 
   CAML_EV_COUNTER(EV_C_MINOR_ALLOCATED, minor_allocated_bytes);
+  CAML_EV_COUNTER(EV_C_MINOR_HEAP_SIZE_WORDS,
+                  (uintnat)Caml_state->minor_heap_wsz);
```

Comment:
It is not obvious to me why you want to track the size of the minor heap as an event counter, and why you would emit the event on each minor collection, while this value rarely (if ever) changes. Wouldn't it make more sense to collect the event in the code that changes the minor heap size? Or where you intending to track something else?

Reply:
The motivating use-case here is we've been considering an analyse subcommand for `olly` in runtime-events-tools that you can connect to an already running OCaml 5 program. It would grab a snapshot of GC/runtime behaviour and then use some rules to identify problems and suggestion solutions. 

Some examples might be consistently very high promotion rates from the minor heap, which indicates it may be too small. Another is periods of time waiting to enter a stop-the-world indicating there's some C code that needs to poll more frequently or a safepoints issue. 

At the moment the problem is if you connect (via runtime events) to an already running program you have no way of knowing how big the minor heap is.

Reply:
(I don't have the details in memory anymore so this may be naive/wrong, but:) from your description it sounds like the current "runtime event" interaction is entirely passive on the listener side, you just take what you get. What you would need for your use-case is to be able to query the working program for its Gc configuration, and there is no such "query" capability in the runtime-event interaction protocol. So your workaround is to broadcast this particular configuration point often enough that a pure listener would get it quickly enough.

Thoughts:
- Should you in fact consider having an active way for monitors to query configuration? If I remember correctly, ocp-memprof would inject into monitored programs some logic to dump its heap description on reception of a particular signal. It would make sense to dump GC-configuration parameters on reception of some signal, or to include this capability in the event-runtime-specific protocol. (Maybe we don't want to rely on signals at all in a multi-domain world.)
- Or, if we want to stick to the "passive listener" approach to simplify the system, should we more systematically broadcast all GC configuration parameters, when and how should we do this?

Reply:
(My gut feeling is that needs a bit more design work and probably a more-encompassing solution than the current one-time choice of broadcasting one bit of GC configuration at an unintuitive time, and I would remove that part from the PR for now or at least stick to the intuitive place to have it, namely when the value changes.)

Reply:
Yes, your understanding of the design is correct and runtime events is passive for listeners. This is deliberate, it isolates the consumer from the running program which brings security if  you are monitoring production applications, there's only one file the consumer needs access to. It also means there's no interaction between consumers when there are multiple.

You could extend the runtime events on-disk format to also include a table of parameters that don't change often but this brings a lot more complexity to save a few bytes emitted every minor collection.

I wouldn't be opposed to broadcasting other GC configuration options if there are motivating use cases for doing so.

> My gut feeling is that needs a bit more design work and probably a more-encompassing solution than the current one-time choice of broadcasting one bit of GC configuration at an unintuitive time

It's deliberately just after the point where we have emitted allocated and promoted words in the minor heap, which is exactly where you need the most up to date value for calculating promotion rates.

> or at least stick to the intuitive place to have it, namely when the value changes

Only having the minor heap size when it changes would rule out being able to connect and analyse a runnning program, I think this would be a shame.

Reply:
> It's deliberately just after the point where we have emitted allocated and promoted words in the minor heap, which is exactly where you need the most up to date value for calculating promotion rates.

What I hear when you say this is: "this unintuitive choice covers just the one exact use-case I have thought about today, I will come back tomorrow change the upstream code when I have a different idea to use my out-of-tree tool". Which is, okay, one possible way to do things, but not my preference as stated above. I think you should think about broadcasting all parameters at the place where it makes sense, probably in a different PR, and not add undocumented nonsensical stuff in the runtime code, even if it is a minor thing.

Reply:
Ok. Fair enough. I'll remove the minor heap size from this PR and follow up with two other PRs, one that adds documentation to the mli and another that emits all GC parameters (maybe at the end of a major cycle). We can continue this discussion on the latter.

For what it's worth I'd be quite happy to contribute runtime-events-tools upstream but until we've finished iterating on it and are sure people find it useful, I don't think it would be a good use of compiler maintainer time.

Reply:
I'm not trying to suggest that the event-tools tooling should be upstreamed, I think that having a minimal part in-tree and the rest out-of-tree is a reasonable design. I was trying to say that I think we should try to design the in-the-runtime part carefully, to maximize utility for out-of-the-runtime consumers (your current tools or other tools in the future, in-tree or out-of-tree) and avoid needing too many changes to the runtime code. I think we should aim for a separation of concern where the capabilities of the analysis code can evolve without requiring never-ending modifications to the runtime code. The only way I know to do this is to think a bit systematically about the interface between the in-the-runtime and out-of-the-runtime part.

Reply:
I've snipped this out of the PR and will follow up with a more general emitting of GC params.